tlv.go

```diff
@@ -0,0 +1,153 @@
+// Type-Length-Value splitting and parsing for proxy protocol V2
+// See spec https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt sections 2.2 to 2.7 and
+// Amazon's extension https://docs.aws.amazon.com/elasticloadbalancing/latest/network/load-balancer-target-groups.html#proxy-protocol
+
+package proxyproto
+
+import (
+	"encoding/binary"
+	"errors"
+	"io"
+	"io/ioutil"
+)
+
+const (
+	// Section 2.2
+	PP2_TYPE_ALPN           PP2Type = 0x01
+	PP2_TYPE_AUTHORITY              = 0x02
+	PP2_TYPE_CRC32C                 = 0x03
+	PP2_TYPE_NOOP                   = 0x04
+	PP2_TYPE_SSL                    = 0x20
+	PP2_SUBTYPE_SSL_VERSION         = 0x21
+	PP2_SUBTYPE_SSL_CN              = 0x22
+	PP2_SUBTYPE_SSL_CIPHER          = 0x23
+	PP2_SUBTYPE_SSL_SIG_ALG         = 0x24
+	PP2_SUBTYPE_SSL_KEY_ALG         = 0x25
+	PP2_TYPE_NETNS                  = 0x30
+	// Section 2.2.7, reserved types
+	PP2_TYPE_MIN_CUSTOM     = 0xE0
+	PP2_TYPE_MAX_CUSTOM     = 0xEF
+	PP2_TYPE_MIN_EXPERIMENT = 0xF0
+	PP2_TYPE_MAX_EXPERIMENT = 0xF7
+	PP2_TYPE_MIN_FUTURE     = 0xF8
+	PP2_TYPE_MAX_FUTURE     = 0xFF
+	// Amazon's extension
+	PP2_TYPE_AWS            = 0xEA
+	PP2_SUBTYPE_AWS_VPCE_ID = 0x01
+)
+
+var (
+	ErrTruncatedTLV    = errors.New("Truncated TLV")
+	ErrMalformedTLV    = errors.New("Malformed TLV Value")
+	ErrIncompatibleTLV = errors.New("Incompatible TLV type")
+)
+
+// PP2Type is the proxy protocol v2 type
+type PP2Type byte
+
+// TLV is a uninterpreted Type-Length-Value for V2 protocol, see section 2.2
+type TLV struct {
+	Type   PP2Type
+	Length int
+	Value  []byte
+}
+
+// readTLVs reads the Type-Length-Value vector, returns the vector or an error.  Reads until EOF or an error occurs.
+func readTLVs(reader io.Reader) ([]TLV, error) {
+	rest, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return nil, err
+	}
+	var tlvs []TLV
+	for i := 0; i < len(rest); {
+		tlv := TLV{
+			Type: PP2Type(rest[i]),
+		}
+		tlv.Length = int(binary.BigEndian.Uint16(rest[i+1 : i+3])) // Max length = 65K
+		i += 3
+		if i+tlv.Length > len(rest) {
+			return nil, ErrTruncatedTLV
+		}
+		// Ignore no-op padding
+		if tlv.Type != PP2_TYPE_NOOP {
+			tlv.Value = make([]byte, tlv.Length)
+			copy(tlv.Value, rest[i:i+tlv.Length])
+		}
+		i += tlv.Length
+		tlvs = append(tlvs, tlv)
+	}
+	return tlvs, nil
+}
+
+// AWSVPCType is true if the TLV is an AWS extension with VPC subtype
+func (t TLV) AWSVPCType() bool {
+	return t.Type.AWS() && t.Length >= 1 && t.Value[0] == PP2_SUBTYPE_AWS_VPCE_ID
+}
+
+// AWSVPCID returns the vpc-id of an AWS VPC extension TLV or errors with ErrIncompatibleTLV or ErrMalformedTLV if
+// it's the wrong TLV type or has a malformed VPC ID (containing chars other than 0-9, a-z, -)
+func (t TLV) AWSVPCID() (string, error) {
+	if !t.AWSVPCType() {
+		return "", ErrIncompatibleTLV
+	}
+	for _, c := range t.Value[1:] {
+		if !(c == byte('-') || (c >= byte('0') && c <= byte('9')) || (c >= byte('a') && c <= byte('z'))) {
```

Comment:
Are upper case chars not valid? I haven't been able to find the documentation for it.

Would it make sense to have regex for this?

Reply:
Yeah there's no docs for the format of amazon resource ids, the defacto standard is `a-z0-9-`, I've changed to a regex that allows caps as well.